src/Microsoft.Health.Fhir.Core/Features/Search/SearchModifier.cs

```diff
@@ -0,0 +1,77 @@
+﻿// -------------------------------------------------------------------------------------------------
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License (MIT). See LICENSE in the repo root for license information.
+// -------------------------------------------------------------------------------------------------
+
+using System;
+using System.Diagnostics.CodeAnalysis;
+using EnsureThat;
+using Microsoft.Health.Fhir.ValueSets;
+
+namespace Microsoft.Health.Fhir.Core.Features.Search
+{
+    /// <summary>
+    /// Defines a search modifier applied to search parameter
+    /// </summary>
+    public sealed class SearchModifier : IEquatable<SearchModifier>
+    {
+        /// <summary>
+        /// Initializes a new instance of the <see cref="SearchModifier"/> class.
+        /// </summary>
+        /// <param name="searchModifierCode"><see cref="SearchModifierCode"/></param>
+        /// <param name="resourceType">Resource type used to constrain abstract modifier code (e.g. <see cref="SearchModifierCode.Type"/>)</param>
+        public SearchModifier(SearchModifierCode searchModifierCode, string resourceType = null)
+        {
+            EnsureArg.EnumIsDefined(searchModifierCode, nameof(searchModifierCode));
```

Comment:
I would drop this check (a simple BenchmarkDotNet test where I comment it out reveals its overhead):

|         Method |       Mean |     Error |    StdDev |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|--------------- |-----------:|----------:|----------:|-------:|------:|------:|----------:|
| WithValidation | 193.409 ns | 3.6804 ns | 5.1595 ns | 0.0105 |     - |     - |      56 B |
|   NoValidation |   7.225 ns | 0.2155 ns | 0.2016 ns | 0.0061 |     - |     - |      32 B |

Reply:
Or make it debug-only.